### PR TITLE
Implemented ScoringHub

### PR DESCRIPTION
**Pull request by u-hubar**

### CODE REVIEW

File: contracts/ScoringHub.sol [hunk 1]
```diff
@@ -0,0 +1,42 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import { IScoringFunction } from "./interface/ScoringFunction.sol";
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+
+
+contract ScroingHub is Ownable {
```
Comment (branarakic) [inline contracts/ScoringHub.sol:9]: I'd rename the contract to ShardScoring, avoiding the word Hub (generally I think we should avoid it to lower the chance of confusing with the main Hub.sol)

File: contracts/ScoringHub.sol [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        returns (uint32)
+    {
+        address scoringContractAddress = functions[scoringFunctionId];
+
+        require(scoringContractAddress != address(0), "Scoring function doesn't exist!");
+
+        IScoringFunction scoringFunction = IScoringFunction(scoringContractAddress);
+        uint256 distance = calculateXORDistance(hashingFunctionId, nodeId, keyword);
+
+        return calculateScore(distance, stake, a, b);
```
Comment (branarakic) [inline contracts/ScoringHub.sol:32]: a and b are not necessary here - they should be part of the scoringContract implementation

File: contracts/interface/ScoringFunction.sol [hunk 3]
```diff
@@ -0,0 +1,10 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+interface IScoringFunction {
+    function calculateScore(uint256 distance, uint96 stake)
+        external returns (uint32);
+    function calculateXORDistance(uint8 hashingFunctionId, bytes memory nodeId, bytes memory keyword)
```
Comment (branarakic) [inline contracts/interface/ScoringFunction.sol:8]: I'd rename to calculateDistance as future distance calculations might not necessarily be an implementation of XOR (for every algorithm)

File: contracts/scoring/dsf.sol [hunk 4]
```diff
@@ -0,0 +1,35 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import { IScoringFunction } from "../interface/ScoringFunction.sol";
+import { HashingHub } from "../HashingHub.sol";
+import { Hub } from "../Hub.sol";
+
+contract DSF is IScoringFunction {
```
Comment (branarakic) [inline contracts/scoring/dsf.sol:9]: Curious, why DSF?

File: contracts/scoring/dsf.sol [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    Hub public hub;
+
+    constructor(address hubAddress) {
+        require(hubAddress != address(0));
+        hub = Hub(hubAddress);
+    }
+
+    function calculateScore(uint256 distance, uint96 stake, uint32 a, uint32 b)
```
Comment (branarakic) [inline contracts/scoring/dsf.sol:17]: in this case "a" and "b" should be variables in the contract, which should also have their respective setters (protected by onlyOwner)

File: contracts/scoring/dsf.sol [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        public
+        returns (uint32)
+    {
+        // VERIFY: uint256 -> uint32 casting
+        return uint32((a * stake) / (b * distance));
```
Comment (branarakic) [inline contracts/scoring/dsf.sol:22]: Should be parametrized to essentially approach a polynomial form. The additional parameters needed, apart from "a" and "b" coefficients, are exponents for stake and distance, and constants in both the numerator and denominator  (all of which should also have their setters)

**Review by branarakic — CHANGES_REQUESTED** [review]

File: contracts/ScoringHub.sol (hunk 1 above)
Comment (u-hubar) [reply contracts/ScoringHub.sol:9]: That's a good point. Maybe ScoringProxy?

File: contracts/scoring/dsf.sol (hunk 4 above)
Comment (u-hubar) [reply contracts/scoring/dsf.sol:9]: > Curious, why DSF?

Default Scoring Function. Didn't come up with anything more original 😅

File: contracts/interface/ScoringFunction.sol (hunk 3 above)
Comment (u-hubar) [reply contracts/interface/ScoringFunction.sol:8]: Done

File: contracts/scoring/dsf.sol (hunk 5 above)
Comment (u-hubar) [reply contracts/scoring/dsf.sol:17]: Done

File: contracts/scoring/dsf.sol (hunk 6 above)
Comment (u-hubar) [reply contracts/scoring/dsf.sol:22]: So idea is to have quadratic polynomials division?

$(a * stake^2 + b * stake + c) \over (d * distance^2 + e * distance + f)$

File: contracts/ScoringHub.sol (hunk 2 above)
Comment (u-hubar) [reply contracts/ScoringHub.sol:32]: Done

File: contracts/scoring/dsf.sol (hunk 4 above)
Comment (u-hubar) [reply contracts/scoring/dsf.sol:9]: Changed it to PLDSF = Polynomial Long Division Scoring Function

**Pull request merged**